server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -325,7 +332,7 @@ public boolean removeDataSource(final String dataSource)
           ).bind("dataSource", dataSource).execute()
       );
 
-      dataSources.remove(dataSource);
+      Optional.ofNullable(dataSources).ifPresent(m -> m.remove(dataSource));
```

Comment:
Why do you use `Optional.ofNullable(dataSources).ifPresent()` instead of
```java
if (dataSources != null) {
  ...
}
```
in this PR?

Reply:
It was because `dataSources` can become null after being non-null, if `stop()` is called. Since `stop()` could be called at any time, `dataSources` should only be dereferenced one time per method.

Reply:
I'll add a comment about this - the variable looks at first glance like a lazy-initialization, but it's actually something that can transition back and forth between null and nonnull, so it needs to be handled differently.

Reply:
Added in https://github.com/apache/incubator-druid/pull/7452.